### ISSUE TITLE
EKS forcibly makes pod<>node NAT mappings endpoint-dependent

### ISSUE THREAD

**Issue by danderson** [tailscale/tailscale#3822]
### What is the issue?

If your service of choice for running Kubernetes clusters is EKS, and you choose to run Tailscale within a pod on the cluster, you will find that Tailscale is now behind a hard NAT, even if the k8s worker VM has a regular AWS public IP.

This is because the EKS CNI sets the `random-fully` option to the SNAT rule that applies to pod<>internet traffic, as seen in netfilter on an EKS worker node:

```
> iptables-legacy -t nat -L KUBE-POSTROUTING -n -v
 4787  292K RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            mark match ! 0x4000/0x4000
    0     0 MARK       all  --  *      *       0.0.0.0/0            0.0.0.0/0            MARK xor 0x4000
    0     0 MASQUERADE  all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* kubernetes service traffic requiring SNAT */ random-fully
```

random-fully makes all the NAT mappings endpoint-dependent, aka a "hard NAT" in our terminology.

I don't know why EKS chose to do this, it's not a standard configuration for Kubernetes, and I can't think of any value it provides.

This bug is mostly for us to document this, since it's a departure from all other k8s services I've seen. A workaround is to run in userspace networking mode and with hostNetwork=true on the pod, which will let Tailscale speak without the pod NAT (but we have to be in userspace mode, otherwise the EKS CNI and we fight each other and break networking completely).

We should also try to reach out to the folks developing the EKS CNI plugin, to understand their reasons for setting this flag - and ideally, encourage them to remove it, so that the NAT behaves like regular linux NAT.

### Steps to reproduce

1. Create an EKS cluster. Settings don't matter, but make sure that the worker VMs have public IPs (i.e. don't put them behind an AWS Managed NAT gateway).
2. Run tailscaled in a pod on the cluster, in userspace mode, with an authkey. Get it all authorized and connected to your tailnet.
3. `tailscale ping` from somewhere else to the EKS pod.
4. Collect tcpdump traffic to/from the EKS pod, on the EKS host and the remote machine.

You'll find in the tcpdumps that the mapped source IP for traffic varies based on the destination ip:port, even though it all originates from a single socket within the container.

### OS

Other

### OS version

AWS EKS, all versions

### Tailscale version

1.20.2

**Comment by danderson** [comment]
According to https://github.com/aws/amazon-vpc-cni-k8s , this is configurable, by passing the envvar AWS_VPC_K8S_CNI_RANDOMIZESNAT=none to the EKS CNI DaemonSet. This will make it use Linux's default NAT behavior, which we can traverse just fine.

I'd still like to figure out why EKS CNI folks chose to default to full randomization, and ideally persuade them to change it back, but in the meantime, if EKS permits reconfiguration of the CNI plugin, that's a viable fix.

**Comment by jazzdan** [comment]
I did some git spelunking and found [this PR](https://github.com/aws/amazon-vpc-cni-k8s/pull/246) which seems to implement this behavior, with the following justification:

>When making a large number of outgoing connections to the same endpoint there is a known race condition with SNAT rules that means that two connections can be assigned the same outgoing port from the ephemeral range. This results in connection traffic (for both the client and server) being unable to track both connections because they will have the same Source IP Address, Source Port, Destination IP Address, and Destination Port, resulting in one process being "orphaned".
>
>To prevent this edge case from occurring you can enable the --random flag within the SNAT rule which will select a random (unused) port from the ephemeral range rather than assigning them sequentially.

**Comment by DentonGentry** [comment]
Cover up a lack of locking by making it 2^15 less likely to occur and therefore deniable.

**Comment by anguslees** [comment]
Yep. Just about every k8s platform did a similar fix at the same time, so I expect this networking situation is more widespread than Dave remembers.

For EKS CNI, there are two possible actions folks can do here: disable random-fully (possibly exposing yourself to the kernel bug), or disable the node-based SNAT entirely.  Disabling node-based SNAT is fine on private nodes (behind NatGW) but necessary on public nodes (behind IGW).

See EXTERNALSNAT in the EKS CNI docs and https://docs.aws.amazon.com/eks/latest/userguide/external-snat.html

**Comment by bradfitz** [comment]
Does anybody know if the Linux bug itself was ever fixed?

**Comment by mayakacz** [comment]
So to recap, it sounds like what we want to do is to document, in our docs on Kubernetes at https://tailscale.com/kb/1185/kubernetes/, that:
* If you are running Tailscale in Kubernetes as a sidecar in EKS, then you will be behind hard NAT. (You may also have this issue on other hosted Kubernetes platforms.)
* To fix this, there are several options:
  1. Run Tailscale as a userspace sidecar (instructions already in that doc)
  1. Disable random-fully SNAT, since you can customize the EKS CNI. This puts you at risk of hitting a bug with SNAT. To do this, pass the envvar AWS_VPC_K8S_CNI_RANDOMIZESNAT=none to the EKS CNI DaemonSet. Link to https://github.com/aws/amazon-vpc-cni-k8s
  1. Disable node-based SNAT. You can do this by passing the envvar AWS_VPC_K8S_CNI_EXTERNALSNAT=true to the node.

Is that right?

**Comment by DentonGentry** [comment]
Perhaps related, the Kubernetes operator is available now: [#502](https://github.com/tailscale/tailscale/issues/502)

**Comment by aviau** [comment]
Do I understand correctly that this is why I can't manage to get direct connections with my tailscale daemons running in EKS?

So, as an EKS user, there is no way to get direct connections without changing the SNAT rule?


> A workaround is to run in userspace networking mode and with hostNetwork=true on the pod, which will let Tailscale speak without the pod NAT

And the tailscale operator does not do that by default? Can it be configured for that?

**Comment by tbroadley** [comment]
I'm also interested in being able to configure the Kubernetes operator to run a Connector with `hostNetwork=true`. I don't see a way to configure the operator to do this.
